internal/schema/schema.go

```diff
@@ -0,0 +1,343 @@
+	ok := errors.As(principalValidationErr, &principalValidationErrorList)
+	if !ok {
+		principalValidationErrorList = nil
+	}
+
+	var resourceValidationErrorList *ValidationErrorList
+	ok = errors.As(resourceValidationErr, &resourceValidationErrorList)
+	if !ok {
+		resourceValidationErrorList = nil
+	}
+
+	numErrors := len(principalValidationErrorList.Errors) + len(resourceValidationErrorList.Errors)
+	validation := make([]ValidationError, 0, numErrors)
+	messages := make([]string, 0, numErrors)
+
+	for _, validationError := range principalValidationErrorList.Errors {
+		validation = append(validation, validationError)
+		messages = append(messages, fmt.Sprintf("%s: %s", validationError.Path,
+			validationError.Message))
+	}
+
+	for _, validationError := range resourceValidationErrorList.Errors {
+		validation = append(validation, validationError)
+		messages = append(messages, fmt.Sprintf("%s: %s", validationError.Path,
+			validationError.Message))
+	}
+
+	printFn := logging.FromContext(ctx).Error
+	if m.conf.Enforcement == EnforcementWarn {
+		printFn = logging.FromContext(ctx).Warn
+	}
+
+	if len(validation) != 0 {
+		printFn("Schema validation failure", zap.Strings("Errors", messages))
+	}
+
+	for _, validationData := range validation {
+		msg := fmt.Sprintf("Schema validation failed for %s attributes: Path=%s - Message=%s",
+			strings.ToLower(validationData.Source.String()),
+			validationData.Path, validationData.Message)
+		if m.conf.Enforcement == EnforcementWarn {
+			logging.FromContext(ctx).Warn(msg)
+		} else {
+			logging.FromContext(ctx).Error(msg)
+		}
+	}
+
+	if m.conf.Enforcement == EnforcementReject {
+		return MergeValidationErrorLists(principalValidationErrorList, resourceValidationErrorList)
+	} else {
+		return nil
+	}
+}
+
+func (m *Manager) validateInput(inputProperties map[string]*structpb.Value, schemaProperties map[string]string,
+	source schemav1.ValidationError_Source) error {
+	if m.conf.IgnoreExtraFields {
+		return nil
+	}
+
+	var properties = make(map[string]string)
+	m.walkInputProperties("", inputProperties, properties)
+
+	var validationErrors []ValidationError
+	for _, property := range properties {
+		_, ok := schemaProperties[property]
+		if !ok {
+			validationErrors = append(validationErrors, ValidationError{
+				Path:    property,
+				Message: "Unexpected field present in the attributes",
+				Source:  source,
+			})
+		}
+	}
+
+	return &ValidationErrorList{
+		Errors: validationErrors,
+	}
+}
+
+func (m *Manager) validatePrincipal(ctx context.Context,
+	principalAttributes map[string]*structpb.Value) error {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	inputValidationErr := m.validateInput(principalAttributes, m.principalProperties, schemav1.ValidationError_SOURCE_PRINCIPAL)
+	if inputValidationErr != nil && !IsValidationErrorList(inputValidationErr) {
+		return fmt.Errorf("failed to validate the input: %w", inputValidationErr)
+	}
+
+	principalAttrBytes, err := json.Marshal(principalAttributes)
```

Comment:
Would this correctly produce the JSON for `structpb` values? Protobufs have a special JSON encoding that the standard library `json` package can't handle (you need to use `protojson`).